### PR TITLE
Working decentralized exchange written in ink!

### PR DESCRIPTION
**Pull request by TomaszWaszczyk**

### CODE REVIEW

**Comment by TomaszWaszczyk** [comment]
@niklasp Hey, please review the PR.

File: tutorials/decentralized-exchange-ink/1-implementation.mdx [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        ZeroLiquidityErr(String),
+        SlippageExceededErr(String),
+        InsufficientAmountErr(String),
+        NonEquivalentValueErr(String),
+        ThresholdNotReachedErr(String),
+        InsufficientLiquidityErr(String),
+    }
+```
+
+In the smart contract is implement some functions that are not strictly needed but helpful for testing, for example a place where a software engineer is able to generate tokens for tests - `faucet_brrr`. To follow K curve in an automated market maker is implemented `get_k` function. To get information about portfolio is implemented `get_information_portfolio` function and for pools information `get_pool_details`.
```
Comment (niklasp) [inline tutorials/decentralized-exchange-ink/1-implementation.mdx:128]: Same as above. Give some reference to the functions you are quoting here. Either links to lines in the github repo or better, copy the relevant parts to this tutorial and put it in cotext. At the moment I am a little lost on what you are talking about.

File: tutorials/decentralized-exchange-ink/1-implementation.mdx [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+git clone git@github.com:TomaszWaszczyk/ink-automated-market-maker.git
+cd ink-automated-market-maker/automated-market-maker
+cargo +nightly build --release
+```
+
+The project uses also `cargo fmt` tool in order to deliver well formatted source code, details how to use the tool you can find in the [link.](https://github.com/rust-lang/rustfmt)
+
+If you are interested in more learning and development of the smart contract one of the option is to implement `proxy pattern`. That is an example of improvement proposal that actually adds real value.
+
+## Github repository
```
Comment (niklasp) [inline tutorials/decentralized-exchange-ink/1-implementation.mdx:150]: This piece of information could be good in the introduction. Something similar to: "this tutorial is based on the github repo ... Go there and explore. The tutorial will then go through explaining the key components of the code" for example.

**Review by niklasp — CHANGES_REQUESTED** [review]
Thanks @TomaszWaszczyk for your tutorial. I added some comments on some lines + suggestions that can help the reader have a better learning experience.

I would also add a link in a paragraph at the end to: https://waszczyk.com/webassembly-first-decentralized-exchange as there are additional information.

Also, how is your planing with the frontend? It would be cool to at least update the tutorial with one section about the frontend as soon as it is done.

Are the smart contracts deployed somewhere where readers can experiment with it? To be an outstanding tutorial that would also be a cool addition alongside example usages.

File: tutorials/decentralized-exchange-ink/1-implementation.mdx [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+```rust
+#[ink(message)]
+pub fn provide_liquidity(
+    &mut self,
+    _amount_token1: Balance,
+    _amount_token2: Balance,
+) -> Result<Balance, Error>
+```
+
+`provide_liquidity` in `ink!` smart contract programming language every public function has to have `#[ink(message)]` attribute. There need to be at least one public method in a smart contract.
```
[the remark this answers is not in the document]
Comment (TomaszWaszczyk) [reply tutorials/decentralized-exchange-ink/1-implementation.mdx]: The word "attribute" I have taken from [ink! documentation](https://use.ink/macros-attributes/message/) to gain for more readability I changed it to "annotation" - I think the word is much more understandable than general attribute.

File: tutorials/decentralized-exchange-ink/1-implementation.mdx (hunk 2 above)
Comment (TomaszWaszczyk) [reply tutorials/decentralized-exchange-ink/1-implementation.mdx:150]: Fully agree. Changed that. I just want to show new people in ink! and/or Rust that here is a fully working project that is usable and compile-able from the first bits of the tutorial. There is very few working examples in the space of ink!.

File: tutorials/decentralized-exchange-ink/1-implementation.mdx [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+`swap` feature is the most complex one in the smart contract and due to a need for estimation firstly of possibility a given swap it is implemented two functions `estimate_swap_token1_for_given_token1` and `estimate_swap_token1_for_given_token2`.
```
[the remark this answers is not in the document]
Comment (TomaszWaszczyk) [reply tutorials/decentralized-exchange-ink/1-implementation.mdx]: agree, hope now is better

File: tutorials/decentralized-exchange-ink/1-implementation.mdx (hunk 1 above)
Comment (TomaszWaszczyk) [reply tutorials/decentralized-exchange-ink/1-implementation.mdx]: agree, hope now is better

**Review by niklasp — APPROVED** [review]
nice! when you are ready, i'd say it is finished. Please update the tutorial as soon as you have a frontend available. A link or better a section would be nice.

send me your ksm address please (again?) in matrix

**Comment by TomaszWaszczyk** [comment]
Thank you for reviewing and approving the article. I would greatly appreciate merging the PR, as I intend to share it on social media. My aim is to shed light on a topic that is relevant to programmers in Poland. Although many of them are highly experienced, there seems to be a tendency towards narrow thinking. Your page carries significant credibility, and I hope that by sharing this article, I can inspire some programmers to develop their own perspectives on reality.

**Pull request merged**